src/if_statements.py

```diff
@@ -353,55 +288,208 @@ def immediate_postdominator(context: Context, start: Node, end: Node) -> Node:
     return postdominators[0]
 
 
-def count_non_postdominated_parents(
-    context: Context, child: Node, curr_end: Node
-) -> int:
+def is_and_statement(next_node: Node, bottom: Node) -> bool:
+    return (
+        not isinstance(next_node, ConditionalNode)
+        or next_node not in bottom.parents
+        # A strange edge-case of our pattern-matching technology:
+        # self-loops match the pattern. Avoiding that...
+        or next_node.is_loop()
+    )
+
+
+def build_conditional_subgraph(
+    context: Context, start: ConditionalNode, end: Node, indent: int
+) -> IfElseStatement:
     """
-    Return the number of parents of "child" for whom "child" is NOT their
-    immediate postdominator. This is useful for finding nodes that would be
-    printed more than once under naive assumptions, i.e. if-conditions that
-    contain multiple predicates in the form of && or ||.
+    Output the subgraph between "start" and "end" at indent level "indent",
+    given that "start" is a ConditionalNode; this program will intelligently
+    output if/else relationships.
     """
-    count = 0
-    for parent in child.parents:
-        if immediate_postdominator(context, parent, curr_end) != child:
-            count += 1
-    # Ideally, either all this node's parents are immediately postdominated by
-    # it, or none of them are. In practice this doesn't always hold, and then
-    # output of && and || may not be correct.
-    if count not in [0, len(child.parents)] and not context.has_warned:
-        context.has_warned = True
-        print(
-            "Warning: confusing control flow, output may have incorrect && "
-            "and || detection. Run with --no-andor to disable detection and "
-            "print gotos instead.\n"
+    if_block_info = start.block.block_info
+    assert isinstance(if_block_info, BlockInfo)
+    assert if_block_info.branch_condition is not None
+
+    # If one of the output edges is the end, it's a "fake" if-statement. That
+    # is, it actually just resides one indentation level above the start node.
```

Comment:
This comment is now even more confusing than before, since the `if start.conditional_edge == end` check is gone